### PR TITLE
Add Japanese

### PR DESCRIPTION
**Pull request by dabura667**

### CODE REVIEW

**Comment by dabura667** [comment]
Ping @weilu @dcousens 

Also, https://github.com/dabura667/bip39-1/commit/6514707eacba3e5377e02fa2e80f43685e1f6b46 is actually a bugfix regardless of Japanese, and would cause deviation if you ever supported multi-byte characters.

**Comment by dcousens** [comment]
LGTM. 
If @weilu isn't able to merge in the next day or so,  I'll merge :+1:

**Comment by dcousens** [comment]
Thanks for this @dabura667!

File: index.js [hunk 1]
```diff
@@ -5,9 +5,10 @@ var randomBytes = require('randombytes')
 var unorm = require('unorm')
 
 var DEFAULT_WORDLIST = require('./wordlists/en.json')
+var JAPANESE_WORDLIST = require('./wordlists/ja.json')
 
 function mnemonicToSeed(mnemonic, password) {
-  var mnemonicBuffer = new Buffer(mnemonic, 'utf8')
+  var mnemonicBuffer = new Buffer(unorm.nfkd(mnemonic), 'utf8')
   var saltBuffer = new Buffer(salt(password), 'utf8')
```
Comment (dcousens) [inline index.js]: Maybe `password` should `unorm`'d here, for consistency.

**Comment by dcousens** [comment]
This is good to go,  thanks @dabura667

File: index.js (hunk 1 above)
Comment (dabura667) [inline index.js]: salt has unorm inside it. So I didn't bother.

However, multiple runs through nfkd will always produce the same result, so maybe for consistency we could do that.

I think it's fine as is though.

**Comment by dcousens** [comment]
@dabura667 my only concern is that this adds another 36K to the file size for this module.
Will make a PR for an idea I have.

**Comment by dcousens** [comment]
I wonder if we can do `require('bip39/japanese')` so that way people aren't unnecessarily including all the word lists on the `bip39` import.

**Comment by dcousens** [comment]
Thoughts @dabura667 / @weilu ?

**Comment by dcousens** [comment]
Will wait on response to above before I publish.

File: index.js (hunk 1 above)
Comment (dcousens) [inline index.js]: Moved them together for consistency in https://github.com/bitcoinjs/bip39/commit/b7d3901adc7d9ff2303fd22be032fba7f3f26856

**Comment by dabura667** [comment]
if all references to wordlists and vectors would iteratively add / test for them instead of explicitly naming them, we could add a function that would go into the wordlist folder and delete a wordlist (which would effectively remove all references to it, if all references to all wordlists was just a loop like:

```javascript
var wordlists = {};
for (var file in fs.readdirSync('./wordlists')) {
  wordlists[file.slice(0,2)] = fs.readFileSync(path.join(__dirname, 'wordlists/' + file), 'utf8');
};
module.exports = {
...
wordlists: wordlists,
...
};
```

Same with the vectors.

And then have a function that does:

```javascript
fs.unlinkSync(path.join(__dirname, 'wordlists/' + filename))
```

To delete the wordlist and therefor all references to it.

**Comment by jprichardson** [comment]
What about creating a new package `bip39-jp` or `bip39-japanese` that would just include the language file. If someone then includes `bip39/japanese` it would use that word list?

**Comment by dabura667** [comment]
One thing I worry about the "default english only" method is that it encourages apps to not support other languages, so support for Japanese takes a hit, lowering ux.

Even if the app doesn't have localization in ja, I believe all apps should be able to recover all types of BIP39 phrases. I believe this was the original vision, by making the seed a hash of the wordlist, however in the real world, all wallets reject invalid phrases outright, so in reality all apps must know all wordlists in order to work.

That is why I think including all, and then a way to opt out, would be best.

**Comment by dcousens** [comment]
@dabura667 that is a good point,   however,   if that application doesn't have e JP localization already,  it seems a bit redundant.  The BIP39 seed will still work,  it just won't validate as per the applications localization.

**Comment by dabura667** [comment]
> The BIP39 seed will still work, it just won't validate as per the applications localization.

That's the problem. A valid Japanese phrase should be allowed EVEN IF japanese localization of the app is unavailable. Not to mention many apps have Japanese localizations with no Japanese BIP39.

I would only agree with separating it if you also set english as a separate require('bip39/english')

That way you make developers at least think of what lists to include.

I would be more inclined to agree with separating them if every wallet app didn't always fail recovery when a valid checksum can't be found (iow: need the wordlist)

The best solution would be for wallet developers to allow any phrase at all, and only show a warning when checksum doesn't check. But no one does that, so I think all BIP39 libraries need all wordlists because of this convention.

I understand why electrum decided against using BIP39. Too many self-contradictions.

**Comment by dcousens** [comment]
> The best solution would be for wallet developers to allow any phrase at all, and only show a warning when checksum doesn't check.

That is what I do :+1:

**Comment by dabura667** [comment]
I wish more devs would think like you... but BIP39 is confusing, so I understand why people reject invalid phrases. I think the checksum should have been brute forced to be contained in the hash of the phrase itself... But maybe that would have not been good for Trezor... idk.

**Comment by dabura667** [comment]
Anywho, my opinion on the matter is that I think it should stay, but if you decide to relegate to separate packages etc. you should:

1. Make it simple to add / remove languages.
2. In the readme explain that devs should allow invalid phrases, but just show a warning. (That way phrases created from unsupported wordlists can be used. (This should be bold or something.)

**Comment by dcousens** [comment]
> In the readme explain that devs should allow invalid phrases, but just show a warning. (That way phrases created from unsupported wordlists can be used. (This should be bold or something.)

I think this deserves its own issue,  and should exist regardless.

**Comment by dabura667** [comment]
> I think this deserves its own issue, and should probably exist regardless of what we do here.

I agree @dcousens , also, I just tried compressing using winrar on the highest compression setting and I got the json file down 77% to 7kB-ish.

Perhaps utilizing some compression when storing and reading the wordlists could lower size constraints.

I'm guessing bitcoinjs-lib is trying to reduce number of dependencies and be "lightweight" with its BIP39 module?

bitcore-mnemonic is (with all dependencies) many many megabytes large...

Maybe using https://nodejs.org/docs/v0.10.40/api/zlib.html#zlib_examples will help out, and make things easier to handle?

**Comment by dcousens** [comment]
> I'm guessing bitcoinjs-lib is trying to reduce number of dependencies and be "lightweight" with its BIP39 module?

I wouldn't say that is entirely the motivation,  but an extra 35k is substantial if that is more than 10% of the current build size for a bitcoinjs-lib browserify build.

**Comment by dabura667** [comment]
@dcousens I just tested, and adding the native nodejs zlib will add 60kB to the minified browserified js file, so even though we could save a few kB on the wordlists, it would be negated by importing zlib.

However, knowing this, it tell's me that if we include 3 wordlists or more, then zipping them for storage and unzipping them into memory on loading only will save space. The more wordlists, the more space saved.

**Pull request merged**